lib/plug/cowboy.ex

```diff
@@ -108,11 +111,29 @@ defmodule Plug.Cowboy do
     run(:https, plug, opts, cowboy_options)
   end
 
+  @doc false
+  @spec drain(module(), Keyword.t()) :: :ok | {:error, term}
+  def drain(ref, opts \\ []) do
+    timeout = Keyword.get(opts, :drain_timeout, 1_000)
+
+    with :ok <- :ranch.suspend_listener(ref) do
+      :ranch.wait_for_connections(ref, :==, 0, timeout)
+    end
+  end
+
   @doc """
   Shutdowns the given reference.
+
+  ## Options
+
+    * `:drain_timeout` - Time in ms to wait for connections to drain during shutdown.
+      Defaults to 1000ms.
   """
-  def shutdown(ref) do
-    :cowboy.stop_listener(ref)
+  @spec shutdown(module(), Keyword.t()) :: :ok | {:error, term}
+  def shutdown(ref, opts \\ []) do
```

Comment:
Hrm... now I see the problem with this approach. When we shutdown, we also want to shutdown the supervisor and the drainer, instead of leaving the supervisor still alive. The only way I can think to implement this is by changing the API so it receives the supervisor PID. :( Thoughts?

Reply:
I was under the impression that `shutdown/1` was intended to be used with `http/3` and `https/3` exclusively, because the listener is being managed by cowboy. Whereas when managing the process yourself, you could just issue a process exit. I can think about how to handle the case where shutdown also is used to stop the supervisor.

Reply:
@blatyo oh, I see what you mean. Yeah, in this case, we can make this clear in the documentation, :+1:.